### ISSUE TITLE
Fetching httponly cookies

### ISSUE THREAD

**Issue by fadhlisulaimi** [react-native-community/react-native-webview#897]
Hi,

I am currently developing an app where the users can login to a website via this webview. I need to fetch httponly cookies so that the user is able to use my app and post contents using his session cookies.

The problem is httponly cookies can only be fetched from the server due to security reasons, in the case only from native iOS and android which is NSHTTPCookieStorage (if i'm correct) and Cookie Manager respectively.

There is another third party package, react-native-cookies which does this but it has been abandoned by the owner and it doesnt fetch httponly cookies, when i tried it with iOS. I'm not sure if it's because of the package not directly attaching to webview, or because of iOS updated security policies. I've looked at it's implementation, it is using NSHTTPCookieStorage in iOS and it's called using native modules, so i'm not sure what's wrong here.

I am wondering if you can include fetching of cookies as well in the webview, especially httponly cookies as it will definitely help a lot of developers out there who creates a similar app as myself. 

Regards

**Comment by taoqf** [comment]
Why you close this issue? @fadhlisulaimi

**Comment by fadhlisulaimi** [comment]
@taoqf I have managed to tweak react-native-cookies package and retrieve httponly cookies from there. I have to change the implementation of the native methods in iOS. I am using WKHTTPCookieStore and WKWebsiteDataStore instead. From there I loop through it and get each of the NSHTTPCookie.

**Comment by rbrcsk** [comment]
Can you share your implementation @fadhlisulaimi ?

**Comment by natcohen** [comment]
@fadhlisulaimi I am also interested to know how you did it!
\
Could you please share the instructions/code?

**Comment by wilson0x4d** [comment]
just my 0.02 ...

1. **cookies "set" by a client are not http-only. by-design it should only be possible to send a cookie as http-only that was set by the (originating) http host/server.** this is a matter of security defined by the protocol and to be enforced by implementations, otherwise, http-only cookies would not exist at all. if you discover a service which accepts a client-set cookie that "should be" http-only then you have discovered a potential vulnerability in the target service: services rely on http-only cookies as a means of securing cookies from malicious scripts/users (abuse.)

2. **http-only cookies should not be readable by client/application code, either.** while you may be able to "hack" the cookie out of a native component via external means it would be inappropriate to do so and-also qualifies as exploitation when done to access secure artifacts created/owned by a third party, such as an auth token, since you are circumventing the security protocols expected to be in effect by the third party. I would not expect such a misuse to pass something like appstore validation except that it may have gone unnoticed.

TL;DR? there _shouldn't_ be a way to read/write http-only cookies from within an application. while doing so may seem useful to a small percentage of hackers it would be inappropriate _as a community_ to make it trivially possible from _within_ react-native applications.  if one "necessarily must" make this possible they would need to do as fadhli has done and modify libraries themselves, but **such modifications should remain unacceptable/inappropriate to the community** as they would make it trivially possible to steal secrets from unsuspecting users.

**Comment by thobiasn** [comment]
@wilson0x4d I absolutely agree, but what solution do we go to then? I am using a httponly cookie setup to authenticate users through a website. I also have an app that ideally should use the same auth endpoint. If I create a new endpoint that returns a token, I'm open to XSS again. If I use the httponly cookie as is, it wont persist through app closures.

**Comment by fadhlisulaimi** [comment]
Basically, due to security concerns...httponly cookies needs to be protected thus it can't be directly retrieved from javascript due to xss or script injection. It can only be retrieved from server or in this case native code.

Thus u need to call httponly cookies from native (android/ios) using cookiemanager (android) and WKHttpCookieStore(ios). Then call the native methods u created from react-native and get the cookie information u need.

**Comment by ansh** [comment]
@fadhlisulaimi can you drop a link to your custom native code?

**Comment by ansh** [comment]
Figured this out by simply using react-native-cookies it is possible

**Comment by albert17** [comment]
@ansh How did you solve this?

**Comment by PaulFlatlooker** [comment]
It should be possible to read an httponly cookie by doing a fetch request like shown here:
https://github.com/react-native-webview/react-native-webview/issues/207#issuecomment-498605263

However I am not sure how to transfer this cookie to the webview afterwards

**Comment by choijiho0021** [comment]
@wilson0x4d 

So how do I keep my login when I shut down the app and turn it on?

**Comment by choijiho0021** [comment]
I read the solutions mentioned above, mostly the same as the library code for https://github.com/react-native-cookies/cookies 
So I checked again in that library,
Turn off Network Inspect on the React Debugger to work.
Why does this happen?

**Comment by johannbuscail** [comment]
Has anyone found a solution ?

**Comment by wilson0x4d** [comment]
> [@wilson0x4d](https://github.com/wilson0x4d) I absolutely agree, but what solution do we go to then? I am using a httponly cookie setup to authenticate users through a website. I also have an app that ideally should use the same auth endpoint. If I create a new endpoint that returns a token, I'm open to XSS again. If I use the httponly cookie as is, it wont persist through app closures.

Actually, you should only be using the cookie to "authorize" requests, not "authenticate" the user. Something else should be used to "authenticate" the user, and then the http-only cookie used to "authorize" requests for the remainder of the session (or until the server invalidates the authorization token.)

> [@PaulFlatlooker](https://github.com/PaulFlatlooker) This doesn't work if httpOnly is set. When httpOnly is set, set-header information cannot be inquired in the response header.
> 
> [@wilson0x4d](https://github.com/wilson0x4d) So how do I keep my login when I shut down the app and turn it on?

Typically you re-authenticate the app when it starts up, establishing another secure (httpOnly) token/cookie to authorize requests over the new session.  I like to store a time-sensitive token on the device that is acquired via a credentialed login (user+pass, SMS code + deviceid, etc/etc) and then use that token for authentication, the server then emits an httpOnly cookie for server-side authorization. The token given to the client contains claims about device, user, and when the token expires forcing the user to authenticate again.

This is not a react-specific problem/pattern, this is how HTTP works "by design" to ensure rogue apps (initially browsers) are not able to trivially hijack and replay server-side cookies.

Hope this helps people do things The Right Way(tm). Good luck!
